torchaudio/prototype/datasets/bucketize_sampler.py

```diff
@@ -0,0 +1,141 @@
+        shuffle: bool = True,
+    ) -> None:
+        if max_len is None:
+            max_len = max(lengths)
+
+        if not (0 <= min_len <= max_len):
+            raise AssertionError("``min_len`` should be non-negative and smaller than ``max_len``")
+        if max_token_count is not None and batch_size is not None:
+            raise AssertionError("The ``max_token_count`` and ``batch_size`` can't be both set.")
+        if max_token_count is None and batch_size is None:
+            raise AssertionError("One of ``max_token_count`` or ``batch_size`` must be set.")
+        # Filter out samples which are outside the bounds of [min_len, max_len]
+        filtered_length_idx = [(length, i) for i, length in enumerate(lengths) if min_len <= length <= max_len]
+        if len(filtered_length_idx) == 0:
+            raise AssertionError("``lengths`` cannot be empty after filtering.")
+        # sort to minimize gap when bucketizing.
+        sorted_filtered_length_idx = sorted(filtered_length_idx, key=lambda x: x[0])
+        self.lengths = [e[0] for e in sorted_filtered_length_idx]
+        self.indices = [e[1] for e in sorted_filtered_length_idx]
+        self.index2iter = {e: i for i, e in enumerate(self.indices)}
+        self.max_token_count = max_token_count
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+        self.buckets = self._get_buckets(self.lengths, self.indices, num_buckets, min_len, max_len)
+
+    def _get_buckets(
+        self, lengths: List[int], indices: List[int], num_buckets: int, min_len: int, max_len: int
+    ) -> Dict[int, Tensor]:
+        """Generate buckets based on the dataset.
+        Args:
+            lengths (List[int]): The lengths of the samples in the dataset.
+            indices (List[int]): The indices of the samples in the original dataset.
+            num_buckets (int): The number of buckets.
+            min_len (int): The lower bound of the evenly spaced length intervals to determine bucket width.
+            max_len (int): The upper bound of the evenly spaced length intervals to determine bucket width.
+
+        Returns:
+            (dict[int, Tensor]): A dictionary in which the key is the bucket index, the value is
+                the Tensor of corresponding sample indices.
+        """
+        buckets = {}
+
+        boundaries = [min_len - 1]
+        interval = (max_len - min_len) // num_buckets
+        for i in range(1, num_buckets):
+            boundaries.append(min_len + i * interval)
+        boundaries.append(max_len + 1)
+        bucket_ids = torch.bucketize(torch.tensor(lengths), torch.tensor(boundaries))
+        for i in indices:
+            bucket_id = bucket_ids[self.index2iter[i]]
+            if bucket_id in buckets:
+                buckets[bucket_id].append(i)
+            else:
+                buckets[bucket_id] = [i]
+        for k in buckets:
+            buckets[k] = torch.as_tensor(buckets[k], dtype=torch.int)
+            if self.shuffle:
```

Comment:
as the caller, i might expect samples to be reshuffled whenever i request a new iterator, rather than shuffled just once during initialization. this behavior would also match that of the default samplers. what do you think?

Reply:
Good point, I will move the shuffle in ``__iter__`` method.